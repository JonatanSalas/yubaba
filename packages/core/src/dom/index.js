import { calculateHypotenuse } from 'math';

export function calculateElementSize (element) {
  return {
    width: element.clientWidth,
    height: element.clientHeight,
  };
}

export function calculateElementLocation (element) {
  const rect = element.getBoundingClientRect();

  return {
    left: rect.left + document.body.clientLeft,
    top: rect.top + document.body.clientTop,
  };
}

export function calculateElementCircumcircle (element) {
  const size = calculateElementSize(element);
  const diameter = calculateHypotenuse(size);
  const location = calculateElementLocation(element);

  return {
    ...location,
    diameter,
  };
}

export function calculateWindowCentre () {
  return {
    left: Math.ceil(window.innerWidth / 2),
    top: Math.ceil(window.innerHeight / 2),
  };
}

/* eslint no-param-reassign:0 */
export function applyStyles (element, styles) {
  Object.keys(styles).forEach((key) => {
    element.style[key] = styles[key];
  });
}

export function calculateElementCenterInViewport (element) {
  const location = calculateElementLocation(element);
  const size = calculateElementSize(element);

  return {
    top: location.top + Math.ceil(size.width / 2),
    left: location.left - Math.ceil(size.height / 2),
  };
}

export function createElement (styles, { parentElement = document.body, cloneFrom }) {
  const newElement = document.createElement('div');
  const innerElement = (cloneFrom && cloneFrom.cloneNode(true));
  if (innerElement) {
    applyStyles(innerElement, {
      margin: 0,
    });

    innerElement.id = '';
    newElement.appendChild(innerElement);
  }

  applyStyles(newElement, {
    ...styles,
  });

  parentElement.appendChild(newElement);
  return newElement;
}
