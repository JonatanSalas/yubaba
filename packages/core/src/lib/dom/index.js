import { calculateHypotenuse } from '../math';

export function calculateElementSize (element) {
  return {
    width: element.clientWidth,
    height: element.clientHeight,
  };
}

export function calculateElementLocation (element/* , calculateViewportScrollOffset*/) {
  // const rect = element.getBoundingClientRect();
  // const offsetY = calculateViewportScrollOffset ? window.scrollY : 0;
  // const offsetX = calculateViewportScrollOffset ? window.scrollX : 0;

  // return {
  //   left: element.offsetLeft, // rect.left + offsetX,
  //   top: element.offsetTop, // rect.top + offsetY,
  // };

  const rect = element.getBoundingClientRect();
  const scrollTop = document.documentElement.scrollTop
    ? document.documentElement.scrollTop
    : document.body.scrollTop;

  const scrollLeft = document.documentElement.scrollLeft
    ? document.documentElement.scrollLeft
    : document.body.scrollLeft;

  return {
    left: rect.left + scrollLeft,
    top: rect.top + scrollTop,
  };
}

export function calculateElementCircumcircle (element) {
  const size = calculateElementSize(element);
  const diameter = calculateHypotenuse(size);
  const location = calculateElementLocation(element);

  return {
    ...location,
    diameter,
  };
}

export function calculateWindowCentre () {
  return {
    left: Math.ceil(window.innerWidth / 2),
    top: Math.ceil(window.innerHeight / 2),
  };
}

const addPxIfNeeded = (key, value) => {
  switch (key) {
    case 'width':
    case 'height':
    case 'top':
    case 'left':
      return typeof value === 'number' ? `${value}px` : value;

    default:
      return value;
  }
};

/* eslint no-param-reassign:0 */
export function applyStyles (element, styles) {
  Object.keys(styles).forEach((key) => {
    element.style[key] = addPxIfNeeded(key, styles[key]);
  });
}

export function calculateElementCenterInViewport (element) {
  const location = calculateElementLocation(element);
  const size = calculateElementSize(element);

  return {
    top: location.top + Math.ceil(size.width / 2),
    left: location.left - Math.ceil(size.height / 2),
  };
}

export function transformTranslate (element, { x, y }) {
  if (!x || !y) {
    return;
  }

  applyStyles(element, {
    transform: `translate3d(${x}px, ${y}px, 0)`,
  });
}

export function transformScale (element, { scale3d, scale, transformOrigin }) {
  if (!scale3d && !scale) {
    return;
  }

  // scale3d takes presedence
  const scaleModifier = scale3d ? `scale3d(${scale3d})` : `scale(${scale})`;

  let transform = element.style.transform;
  if (transform.indexOf('scale') > -1) {
    // replace any prexisting transform with a new one.
    transform = transform.replace(/scale\(.*\)|scale3d\(.*\)/, scaleModifier);
  } else if (transform.length > 0) {
    transform = `${element.style.transform} ${scaleModifier}`;
  } else {
    transform = scaleModifier;
  }

  applyStyles(element, {
    transformOrigin,
    backfaceVisibility: 'hidden',
    webkitBackgroundClip: 'content-box',
    transform,
  });
}

// eslint-disable-next-line
function logStyle (element, styles) {
  return Object.keys(styles).reduce((obj, key) => ({
    ...obj,
    [key]: element.style[key],
  }), {});
}

export function createElement (styles, { parentElement = document.body, cloneFrom } = {}) {
  const newElement = document.createElement('div');
  const innerElement = (cloneFrom && cloneFrom.cloneNode(true));
  if (innerElement) {
    applyStyles(innerElement, {
      margin: 0,
    });

    innerElement.id = '';
    newElement.appendChild(innerElement);
  }

  applyStyles(newElement, {
    ...styles,
  });

  transformScale(newElement, styles);

  parentElement.appendChild(newElement);
  return newElement;
}
