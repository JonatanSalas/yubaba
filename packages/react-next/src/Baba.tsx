import * as React from 'react';

/*
  v1 API

  EXAMPLE ONE - Single move animation.

  Pass ref all the way up to Baba, collecting configuration up.

  // This would move image1 over to image2 when this unmounts.
  <Baba name="image1-to-image2">
    <Move duration={300}>
      {(ref) => <Image innerRef={ref} />}
    </Move>
  </Baba>

  // This would move image2 over to image1 when this unmounts.
  <Baba name="image1-to-image2">
    <Move duration={150}>
      {(ref) => <Image innerRef={ref} />}
    </Move>
  </Baba>

  EXAMPLE TWO - Two animations.

  // 1. move image1 over to image2
  // 2. AT THE SAME TIME circle expand to cover viewport
  <Baba name="image1-to-image2">
    <CircleExpand duration={300} background="#3d7596">
      <Move duration={150}>
        {(ref) => <Image innerRef={ref} />}
      </Move>
    </CircleExpand>
  </Baba>

  // 1. move image1 over to image2
  // 2. WAIT until other transitions are finished, then
  // move shrink a circle from viewport to over image2
  <Baba name="image1-to-image2">
    <CircleExpand duration={300} background="#3d7596" wait>
      <Move duration={150}>
        {(ref) => <Image innerRef={ref} />}
      </Move>
    </CircleExpand>
  </Baba>
*/

interface Props {
  name: string;
}

export default class Baba extends React.Component<Props> {
  render() {
    return null;
  }
}
