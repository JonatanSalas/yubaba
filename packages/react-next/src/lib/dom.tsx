/**
 * @hidden
 */
export function getDocumentScroll() {
  const scrollTop =
    document.documentElement && document.documentElement.scrollTop
      ? document.documentElement.scrollTop
      : document.body && document.body.scrollTop;

  const scrollLeft =
    document.documentElement && document.documentElement.scrollLeft
      ? document.documentElement.scrollLeft
      : document.body && document.body.scrollLeft;

  return {
    scrollTop: scrollTop || 0,
    scrollLeft: scrollLeft || 0,
  };
}

/**
 * @hidden
 */
export interface GetElementSizeLocationOptions {
  useOffsetSize?: boolean;
}

/**
 * @hidden
 */
export interface GetElementSizeLocationReturnValue {
  size: {
    width: number;
    height: number;
  };
  location: {
    left: number;
    top: number;
  };
  raw: {
    rect: ClientRect | DOMRect;
    scrollTop: number;
    scrollLeft: number;
  };
}

/**
 * @hidden
 */
export function getElementSizeLocation(
  element: HTMLElement,
  options: GetElementSizeLocationOptions = {}
): GetElementSizeLocationReturnValue {
  const rect = element.getBoundingClientRect();
  const { scrollLeft, scrollTop } = getDocumentScroll();
  const topOffset = (rect.height - element.offsetHeight) / 2;
  const leftOffset = (rect.width - element.offsetWidth) / 2;

  return {
    size: {
      width: options.useOffsetSize ? element.offsetWidth : rect.width,
      height: options.useOffsetSize ? element.offsetHeight : rect.height,
    },
    location: {
      left: rect.left + scrollLeft + leftOffset,
      top: rect.top + scrollTop + topOffset,
    },
    raw: {
      rect,
      scrollTop,
      scrollLeft,
    },
  };
}

/**
 * @hidden
 */
export function calculateElementCenterInViewport(sizeLocation: GetElementSizeLocationReturnValue) {
  return {
    top: sizeLocation.location.top + Math.ceil(sizeLocation.size.width / 2),
    left: sizeLocation.location.left - Math.ceil(sizeLocation.size.height / 2),
  };
}

/**
 * @hidden
 */
export function calculateWindowCentre() {
  return {
    left: Math.ceil(window.innerWidth / 2),
    top: Math.ceil(window.innerHeight / 2),
  };
}

/**
 * @hidden
 */
export function recalculateLocationFromScroll(
  sizeLocation: GetElementSizeLocationReturnValue
): GetElementSizeLocationReturnValue {
  const { scrollTop, scrollLeft } = getDocumentScroll();
  const scrollTopDiff = scrollTop - sizeLocation.raw.scrollTop;
  const scrollLeftDiff = scrollLeft - sizeLocation.raw.scrollLeft;

  return {
    ...sizeLocation,
    location: {
      top: sizeLocation.location.top + scrollTopDiff,
      left: sizeLocation.location.left + scrollLeftDiff,
    },
  };
}
