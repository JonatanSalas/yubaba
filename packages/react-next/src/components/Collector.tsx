import * as React from 'react';
import { GetElementSizeLocationReturnValue } from '../lib/dom';

export type AnimationCallback = (data: AnimationData) => Promise<any>;
export enum Actions {
  animation = 'animation',
  wait = 'wait',
}
export type Data =
  | {
      action: Actions.animation;
      payload: {
        animate: AnimationCallback;
        beforeAnimate?: AnimationCallback;
        afterAnimate?: AnimationCallback;
        abort?: () => void;
        cleanup: () => void;
      };
    }
  | { action: Actions.wait };
export type SupplyRef = (ref: HTMLElement | null) => void;
export type SupplyRenderChildren = (reactNode: ChildrenAsFunction) => void;
export type SupplyData = (data: Data[]) => void;
export type ChildrenAsFunction = (props: { ref: SupplyRef; style: Style }) => React.ReactNode;

export interface AnimationData {
  caller: React.Component;
  fromTarget: TargetData;
  toTarget: TargetData;
}

export interface Style {
  [key: string]: string | number | undefined;
}

export interface TargetData extends GetElementSizeLocationReturnValue {
  render: ChildrenAsFunction;
}

export interface CommonProps {
  children: ChildrenAsFunction | React.ReactNode;
}

export interface Props extends CommonProps {
  receiveRef?: SupplyRef;
  receiveRenderChildren?: SupplyRenderChildren;
  receiveData?: SupplyData;
  data?: Data;
  style?: Style;
}

export interface Collect {
  ref: SupplyRef;
  data: SupplyData;
  renderChildren: SupplyRenderChildren;
  style: Style;
}

const CollectContext = React.createContext<Collect>();

export default class Collector extends React.Component<Props> {
  render() {
    if (typeof this.props.children !== 'function') {
      return (
        <CollectContext.Consumer>
          {collect => (
            <CollectContext.Provider
              value={{
                ref: ref => {
                  this.props.receiveRef && this.props.receiveRef(ref);
                  collect && collect.ref(ref);
                },
                data: childData => {
                  const data = this.props.data ? [this.props.data].concat(childData) : childData;
                  collect && collect.data(data);
                  this.props.receiveData && this.props.receiveData(childData);
                },
                renderChildren: node => {
                  collect && collect.renderChildren(node);
                  this.props.receiveRenderChildren && this.props.receiveRenderChildren(node);
                },
                style: {
                  ...this.props.style,
                  ...(collect ? collect.style : {}),
                },
              }}
            >
              {this.props.children}
            </CollectContext.Provider>
          )}
        </CollectContext.Consumer>
      );
    }

    return (
      <CollectContext.Consumer>
        {collect => {
          if (typeof this.props.children === 'function') {
            if (collect) {
              const data = this.props.data ? [this.props.data] : [];
              collect.renderChildren(this.props.children);
              collect.data(data);
            }

            if (this.props.receiveRenderChildren) {
              this.props.receiveRenderChildren(this.props.children);
            }

            return this.props.children({
              ref: (ref: HTMLElement) => {
                if (collect) {
                  collect.ref(ref);
                }

                if (this.props.receiveRef) {
                  this.props.receiveRef(ref);
                }
              },
              style: collect ? { ...this.props.style, ...collect.style } : this.props.style || {},
            });
          }

          throw new Error('Children is  guaranteed to be a function. Impossible condition.');
        }}
      </CollectContext.Consumer>
    );
  }
}
