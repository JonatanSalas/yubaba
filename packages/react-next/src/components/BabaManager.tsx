import * as React from 'react';
import { InlineStyles } from './Collector';

export interface BabaManangerProps {
  /**
   * Children as function which passes down style,
   * add this to the elements you want to hide until all child animations have finished.
   */
  children: (props: { style: InlineStyles }) => React.ReactNode;

  /**
   * Optional name to target a specific child `<Baba />`.
   */
  name?: string;
}

/**
 * @hidden
 */
export interface State {
  style: InlineStyles;
}

/**
 * @hidden
 */
export interface InjectedProps {
  context?: BabaManagerContext;
}

/**
 * @hidden
 */
export type OnFinishHandler = (opts: { name: string }) => void;

/**
 * @hidden
 */
export interface BabaManagerContext {
  onFinish: OnFinishHandler;
}

/**
 * @hidden
 */
export const BabaContext = React.createContext<BabaManagerContext>();

/**
 * ## BabaManager
 *
 * Used to hide contents before an animation is complete triggered from a child `<Baba />` component.
 * If there is more than one child `<Baba />` you can use an optional `name` prop which should match the appropriate `<Baba />` component.
 *
 * If it is the initial mount it will immediately be shown.
 *
 * ### Usage
 *
 * ```
 * const MyApp = () => (
 * <BabaManager name="my-anim">
 *    {(props) => (
 *      <div {...props}>
 *        <Baba name="my-anim">
 *          {(props) => <div {...props} />}
 *        </Baba>
 *      </div>
 *    )}
 *   </BabaManager>
 * );
 * ```
 */
export default class BabaManager extends React.Component<BabaManangerProps, State> {
  state: State = {
    style: {
      opacity: 0,
    },
  };

  onFinish: OnFinishHandler = opts => {
    if (this.props.name && opts.name !== this.props.name) {
      return;
    }

    this.setState({
      style: {
        opacity: 1,
      },
    });
  };

  render() {
    return (
      <BabaContext.Provider value={{ onFinish: this.onFinish }}>
        {this.props.children({ style: this.state.style })}
      </BabaContext.Provider>
    );
  }
}

/**
 * @hidden
 */
export const withBabaManagerContext = <T extends InjectedProps>(
  WrappedComponent: React.ComponentType<T>
) => (props: T) => (
  <BabaContext.Consumer>
    {context => <WrappedComponent context={context} {...props} />}
  </BabaContext.Consumer>
);
