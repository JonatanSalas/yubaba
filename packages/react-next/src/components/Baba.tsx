import * as React from 'react';
import Collector, {
  SupplyData,
  SupplyRenderChildren,
  SupplyRef,
  ChildrenAsFunction,
  Data,
  AnimationResult,
} from './Collector';
import { getElementSizeLocation } from '../lib/dom';
import * as childrenStore from '../lib/childrenStore';

type StartAnimation = () => Promise<AnimationResult>;
type AnimationBlock = StartAnimation[];

/*
  v1 API

  EXAMPLE ONE - Single move animation.

  Pass ref all the way up to Baba, collecting configuration up.

  // This would move image1 over to image2 when this unmounts.
  <Baba name="image1-to-image2">
    <Move duration={300}>
      {(ref) => <Image innerRef={ref} />}
    </Move>
  </Baba>

  // This would move image2 over to image1 when this unmounts.
  <Baba name="image1-to-image2">
    <Move duration={150}>
      {(ref) => <Image innerRef={ref} />}
    </Move>
  </Baba>

  EXAMPLE TWO - Two animations with a wait.

  // 1. move image1 over to image2
  // 2. AT THE SAME TIME circle expand to cover viewport
  <Baba name="image1-to-image2">
    <CircleExpand duration={300} background="#3d7596">
      <Move duration={150}>
        {(ref) => <Image innerRef={ref} />}
      </Move>
    </CircleExpand>
  </Baba>

  // 1. move image1 over to image2
  // 2. WAIT until other transitions are finished, then
  // move shrink a circle from viewport to over image2
  <Baba name="image1-to-image2">
    <CircleExpand duration={300} background="#3d7596">
      <Wait>
        <Move duration={150}>
          {(ref) => <Image innerRef={ref} />}
        </Move>
      </Wait>
    </CircleExpand>
  </Baba>
*/

interface Props {
  name: string;
  children: React.ReactNode;
  onFinish?: () => {};
}

interface State {
  shown: boolean;
}

export default class Baba extends React.PureComponent<Props, State> {
  state: State = {
    shown: false,
  };

  element: HTMLElement | null;
  renderChildren: ChildrenAsFunction;
  data: Data[];
  hasStoredBefore: boolean = false;
  cancelClear?: () => void;

  componentDidMount() {
    if (childrenStore.has(this.props.name)) {
      // A child has already been stored, so this is probably the matching pair.
      // Lets execute!
      return this.execute();
    } else {
      // Ok nothing is there yet, show our children and store DOM data for later.
      // We'll be waiting for another <Baba /> instance to mount.
      this.setState({
        shown: true,
      });
      this.store();
    }

    return undefined;
  }

  componentWillUnmount() {
    this.delayedClear();
  }

  componentDidUpdate() {
    if (this.hasStoredBefore) {
      // This instance has stored before, and we've been updated. Let's store DOM data for later.
      this.store();
    }
  }

  delayedClear() {
    const id = setTimeout(() => {
      childrenStore.remove(this.props.name);
    }, 50);

    return () => clearTimeout(id);
  }

  store() {
    childrenStore.set(this.props.name, {
      ...getElementSizeLocation(this.element as HTMLElement),
      element: this.element as HTMLElement,
      render: this.renderChildren,
      data: this.data,
    });

    this.hasStoredBefore = true;
  }

  execute() {
    const fromTarget = childrenStore.get(this.props.name);
    if (fromTarget) {
      const { data, ...target } = fromTarget;

      const blocks = fromTarget.data.reduce<AnimationBlock[]>(
        (arr, data) => {
          switch (data.action) {
            case 'animation': {
              const animate = data.payload;

              // Add to the last block in the array.
              arr[arr.length - 1].push(() =>
                animate({
                  caller: this,
                  fromTarget: target,
                  toTarget: {
                    render: this.renderChildren,
                    ...getElementSizeLocation(this.element as HTMLElement),
                  },
                })
              );

              return arr;
            }

            case 'wait': {
              // Found a wait action, start a new block.
              arr.push([]);
              return arr;
            }

            default: {
              return arr;
            }
          }
        },
        [[]]
      );

      // Trigger each blocks animations, one block at a time.
      return blocks
        .reduce<Promise<AnimationResult>>(
          (promise, block) => promise.then(() => Promise.all(block.map(animate => animate()))),
          Promise.resolve({} as AnimationResult)
        )
        .then(() => {
          // We're finished all the transitions! Show the child element.
          this.setState({
            shown: true,
          });

          // We don't need the previous children now. Now this instance is the new target!
          // Store DOM data for later so when another target is mounted, the data is there.
          this.store();

          if (this.props.onFinish) {
            this.props.onFinish();
          }
        });
    }

    return undefined;
  }

  setRef: SupplyRef = ref => {
    this.element = ref;
  };

  setReactNode: SupplyRenderChildren = renderChildren => {
    this.renderChildren = renderChildren;
  };

  setData: SupplyData = data => {
    this.data = data;
  };

  render() {
    return (
      <Collector
        receiveData={this.setData}
        receiveRenderChildren={this.setReactNode}
        receiveRef={this.setRef}
        style={{
          opacity: this.state.shown ? 1 : 0,
        }}
      >
        {this.props.children}
      </Collector>
    );
  }
}
