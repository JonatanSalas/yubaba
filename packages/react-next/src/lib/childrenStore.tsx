import { GetElementSizeLocationReturnValue } from '../lib/dom';
import { CollectorData, CollectorChildrenAsFunction } from '../components/Collector';

/**
 * @hidden
 */
export interface ChildrenData extends GetElementSizeLocationReturnValue {
  render: CollectorChildrenAsFunction;
  data: CollectorData[];
}

/**
 * @hidden
 */
const children = new Map<string, ChildrenData>();

/**
 * @hidden
 */
export const set = (key: string, value: ChildrenData) => children.set(key, value);

/**
 * @hidden
 */
export const get = (key: string) => children.get(key);

/**
 * @hidden
 */
export const has = (key: string) => children.has(key);

/**
 * @hidden
 */
export const remove = (key: string) => children.delete(key);
