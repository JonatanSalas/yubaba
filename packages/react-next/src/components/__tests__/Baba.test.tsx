import * as React from 'react';
import { shallow, ShallowWrapper } from 'enzyme';
import Baba from '../Baba';
import Collector, { Actions, Data } from '../Collector';
import * as childrenStore from '../../lib/childrenStore';
import * as dom from '../../lib/dom';
import * as BabaManager from '../BabaManager';

jest.mock('../../lib/childrenStore');
jest.mock('../../lib/dom');
jest.mock('../BabaManager', () => ({
  withBabaManagerContext: component => component,
}));

describe('<Baba />', () => {
  afterEach(() => {
    jest.resetAllMocks();
  });

  const stubAllReceivers = (wrapper: ShallowWrapper) => {
    const collector = wrapper.find(Collector);
    const { receiveData, receiveRef, receiveRenderChildren } = collector.props();

    receiveData([]);
    receiveRef(document.createElement('div'));
    receiveRenderChildren(() => <div />);
  };

  const createAnimationPayload = (animate: any): Data => ({
    action: Actions.animation,
    payload: {
      animate,
      prepare: () => Promise.resolve(),
      abort: () => {},
      cleanup: () => {},
    },
  });

  const shallowRender = () => {
    const wrapper = shallow(
      <Baba name="my-animation">
        <div />
      </Baba>,
      {
        disableLifecycleMethods: true,
      }
    );

    stubAllReceivers(wrapper);

    return {
      wrapper,
      mount: () => wrapper.instance().componentDidMount(),
    };
  };

  describe('storing DOM data when unmounting', () => {
    it('should store data in named key', () => {
      const { wrapper } = shallowRender();

      wrapper.unmount();

      expect((childrenStore.set as jest.Mock).mock.calls[0][0]).toEqual('my-animation');
    });

    it('should store position data', () => {
      const { wrapper } = shallowRender();
      const ref = {} as HTMLElement;
      const sizeLocation = { size: {}, location: {} };
      const { receiveRef } = wrapper.find(Collector).props();
      receiveRef(ref);
      (dom.getElementSizeLocation as jest.Mock).mockReturnValueOnce(sizeLocation);

      wrapper.unmount();

      expect((childrenStore.set as jest.Mock).mock.calls[0][1]).toMatchObject({
        ...sizeLocation,
      });
    });

    it('should store ref', () => {
      const { wrapper } = shallowRender();
      const ref = {} as HTMLElement;
      const { receiveRef } = wrapper.find(Collector).props();
      receiveRef(ref);

      wrapper.unmount();

      expect((childrenStore.set as jest.Mock).mock.calls[0][1]).toMatchObject({
        element: ref,
      });
    });

    it('should store react node', () => {
      const { wrapper } = shallowRender();
      const render = () => <div />;
      const { receiveRenderChildren } = wrapper.find(Collector).props();
      receiveRenderChildren(render);

      wrapper.unmount();

      expect((childrenStore.set as jest.Mock).mock.calls[0][1]).toMatchObject({
        render,
      });
    });

    it('should show children immediately', () => {
      const wrapper = shallow(
        <Baba name="my-animation">
          <div />
        </Baba>
      );

      expect(wrapper.find(Collector).props().style).toEqual({ opacity: 1 });
    });
  });

  describe('cleanup', () => {
    it('should clear out child store if a match isnt found in time', () => {
      const { wrapper, mount } = shallowRender();
      stubAllReceivers(wrapper);
      mount();
      jest.useFakeTimers();

      wrapper.unmount();

      jest.runAllTimers();
      jest.useRealTimers();
      expect(childrenStore.remove).toBeCalledWith('my-animation');
    });
  });

  describe('triggering animations', () => {
    const prepare = (data: Data[]) => {
      (childrenStore.has as jest.Mock).mockReturnValue(true);
      (childrenStore.get as jest.Mock).mockReturnValue({ data });
    };

    it('should pass through data to animation', async () => {
      const { wrapper, mount } = shallowRender();
      const animation = jest.fn().mockResolvedValue({});
      const data: Data[] = [createAnimationPayload(animation)];
      prepare(data);

      await mount();

      expect(animation.mock.calls[0][0].caller).toEqual(wrapper.instance());
    });

    it('should wait until animation has finished if a wait was found', async () => {
      const { wrapper, mount } = shallowRender();
      const longAnimation = () => Promise.resolve({});
      const animation = jest.fn().mockResolvedValue({});
      const data: Data[] = [
        createAnimationPayload(longAnimation),
        { action: Actions.wait },
        createAnimationPayload(animation),
      ];
      prepare(data);

      const promise = mount();

      await longAnimation();
      expect(animation).not.toHaveBeenCalled();
      await promise;
      expect(animation).toHaveBeenCalled();
    });

    it('should hide children when starting animating', () => {
      const { wrapper, mount } = shallowRender();
      const animation = jest.fn().mockResolvedValue({});
      const data: Data[] = [createAnimationPayload(animation)];
      prepare(data);

      mount();

      expect(wrapper.find(Collector).props().style).toEqual({ opacity: 0 });
    });

    it('should show children when finished animating', async () => {
      const { wrapper, mount } = shallowRender();
      const animation = jest.fn().mockResolvedValue({});
      const data: Data[] = [createAnimationPayload(animation)];
      prepare(data);

      await mount();
      wrapper.update();

      expect(wrapper.find(Collector).props().style).toEqual({ opacity: 1 });
    });

    it('should store data after finishing so the next target can pick it up', async () => {
      const { wrapper, mount } = shallowRender();
      const animation = jest.fn().mockResolvedValue({});
      const data: Data[] = [createAnimationPayload(animation)];
      prepare(data);
      stubAllReceivers(wrapper);

      await mount();

      expect(childrenStore.set as jest.Mock).toHaveBeenCalled();
    });
  });
});
