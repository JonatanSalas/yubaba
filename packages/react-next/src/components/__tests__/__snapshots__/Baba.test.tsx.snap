// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Baba /> should pass dom data to child animation 1`] = `
Array [
  Object {
    "fromTarget": Object {
      "location": Object {
        "left": 0,
        "top": 0,
      },
      "raw": Object {
        "rect": Object {},
        "scrollLeft": 0,
        "scrollTop": 0,
      },
      "render": [Function],
      "size": Object {
        "height": 200,
        "width": 100,
      },
    },
    "toTarget": Object {
      "location": Object {
        "left": 0,
        "top": 0,
      },
      "raw": Object {
        "rect": Object {},
        "scrollLeft": 0,
        "scrollTop": 0,
      },
      "render": undefined,
      "size": Object {
        "height": 200,
        "width": 100,
      },
    },
  },
]
`;
